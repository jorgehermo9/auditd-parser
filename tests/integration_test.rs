use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
