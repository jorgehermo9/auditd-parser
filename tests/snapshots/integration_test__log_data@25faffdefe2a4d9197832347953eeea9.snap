---
source: tests/integration_test.rs
expression: record
info: "type=USER_ACCT msg=audit(1481077043.052:410): pid=1298 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:sshd_t:s0-s0:c0.c1023 msg='op=PAM:accounting grantors=pam_unix,pam_localuser acct=\"andrew_kroh\" exe=\"/usr/sbin/sshd\" hostname=pool-96-241-146-97.washdc.fios.verizon.net addr=96.241.146.97 terminal=ssh res=success'"
---
{
  "record_type": "USER_ACCT",
  "timestamp": 1481077043052,
  "id": 410,
  "fields": {
    "auid": null,
    "msg": {
      "acct": "andrew_kroh",
      "addr": "96.241.146.97",
      "exe": "/usr/sbin/sshd",
      "grantors": "pam_unix,pam_localuser",
      "hostname": "pool-96-241-146-97.washdc.fios.verizon.net",
      "op": "PAM:accounting",
      "res": "success",
      "terminal": "ssh"
    },
    "pid": "1298",
    "ses": "4294967295",
    "subj": {
      "level": {
        "category": "c0.c1023",
        "sensitivity": "s0-s0"
      },
      "module": "SELinux",
      "role": "system_r",
      "type": "sshd_t",
      "user": "system_u"
    },
    "uid": "root"
  }
}
